partiql-eval/src/eval/expr/operators.rs

```diff
@@ -0,0 +1,327 @@
+use crate::eval::expr::eval_wrapper::{
+    unwrap_args, ArgCheckControlFlow, ArgCheckEvalExpr, ArgChecker, ArgShortCircuit,
+    BinaryValueExpr, DefaultArgChecker, ExecuteEvalExpr, NullArgChecker, PropagateMissing,
+    PropagateNull, TernaryValueExpr, UnaryValueExpr,
+};
+
+use crate::eval::expr::{BindError, BindEvalExpr, EvalExpr};
+use crate::eval::EvalContext;
+
+use partiql_types::{
+    ArrayType, BagType, PartiqlType, StructType, TypeKind, TYPE_ANY, TYPE_BOOL, TYPE_DECIMAL,
+    TYPE_DOUBLE, TYPE_INT, TYPE_REAL,
+};
+use partiql_value::Value::{Boolean, Missing, Null};
+use partiql_value::{BinaryAnd, BinaryOr, NullableEq, NullableOrd, Value};
+
+use std::borrow::{Borrow, Cow};
+use std::fmt::Debug;
+
+use std::marker::PhantomData;
+
+use std::ops::ControlFlow;
+
+/// Represents a literal in (sub)query, e.g. `1` in `a + 1`.
+#[derive(Debug)]
+pub(crate) struct EvalLitExpr {
+    pub(crate) lit: Box<Value>,
+}
+
+impl BindEvalExpr for EvalLitExpr {
+    fn bind<const STRICT: bool>(
+        &self,
+        args: Vec<Box<dyn EvalExpr>>,
+    ) -> Result<Box<dyn EvalExpr>, BindError> {
+        let expr: ArgCheckEvalExpr<
+            STRICT,
+            0,
+            _,
+            DefaultArgChecker<{ STRICT }, PropagateMissing<false>>,
+        > = ArgCheckEvalExpr::new([], unwrap_args(args)?, self.lit.as_ref().clone());
+        Ok(Box::new(expr))
+    }
+}
+
+impl ExecuteEvalExpr<0> for Value {
+    fn evaluate<'a>(
+        &'a self,
+        _args: [Cow<'a, Value>; 0],
+        _ctx: &'a dyn EvalContext,
+    ) -> Cow<'a, Value> {
+        Cow::Borrowed(self)
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub(crate) enum EvalOpUnary {
+    Pos,
+    Neg,
+    Not,
+}
+
+impl BindEvalExpr for EvalOpUnary {
+    fn bind<const STRICT: bool>(
+        &self,
+        args: Vec<Box<dyn EvalExpr>>,
+    ) -> Result<Box<dyn EvalExpr>, BindError> {
+        let any_num = PartiqlType::any_of([TYPE_INT, TYPE_REAL, TYPE_DOUBLE, TYPE_DECIMAL]);
+
+        let unop = |types, f: fn(&Value) -> Value| {
+            UnaryValueExpr::create_typed::<{ STRICT }, _>(types, args, f)
+        };
+
+        match self {
+            EvalOpUnary::Pos => unop([any_num], |operand| operand.clone()),
+            EvalOpUnary::Neg => unop([any_num], |operand| -operand),
+            EvalOpUnary::Not => unop([TYPE_BOOL], |operand| !operand),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub(crate) enum EvalOpBinary {
+    // Logical ops
+    And,
+    Or,
+
+    // Equality ops
+    Eq,
+    Neq,
+    Gt,
+    Gteq,
+    Lt,
+    Lteq,
+
+    // Arithmetic ops
+    Add,
+    Sub,
+    Mul,
+    Div,
+    Mod,
+    Exp,
+
+    // other ops
+    In,
+    Concat,
+}
+
+#[derive(Debug)]
+struct BoolShortCircuitArgChecker<const B: bool, OnMissing: ArgShortCircuit> {
```

Comment:
nit: could add some documentation on what `B` represents

Reply:
Maybe moving to `eval_wrapper`?